openvdb/openvdb/tools/Morphology.h

```diff
@@ -63,540 +59,835 @@ enum NearestNeighbors { NN_FACE = 6, NN_FACE_EDGE = 18, NN_FACE_EDGE_VERTEX = 26
+        const auto range = mManager.getRange(this->getThreaded() ?
+            1 : std::numeric_limits<size_t>::max());
+        tbb::parallel_for(range,
+            [&](const tbb::blocked_range<size_t>& r){
+            for (size_t idx = r.begin(); idx < r.end(); ++idx)
+                masks[idx] = mManager.leaf(idx).getValueMask();
+        });
+    }
+
+public:
+    /// @brief  Node Mask dilation/erosion operations for individual leaf nodes on
+    ///   a given tree. The leaf node may optionally belong to a different tree
+    ///   than the provided accessor, which will have the effect of dilating the
+    ///   leaf node mask into a different tree, or eroding the node mask based
+    ///   on corresponding neighbours in a different tree.
+    struct NodeMaskOp
+    {
+        static const Int32 DIM = static_cast<Int32>(LeafType::DIM);
+        static const Int32 LOG2DIM = static_cast<Int32>(LeafType::LOG2DIM);
+
+        // Select the storage size based off the dimensions of the leaf node
+        using Word = typename std::conditional<LOG2DIM == 3, uint8_t,
+            typename std::conditional<LOG2DIM == 4, uint16_t,
+            typename std::conditional<LOG2DIM == 5, uint32_t,
+            typename std::conditional<LOG2DIM == 6, uint64_t,
+                void>::type>::type>::type>::type;
+
+        static_assert(!std::is_same<Word, void>::value,
+            "Unsupported Node Dimension for node mask dilation/erosion");
+
+        NodeMaskOp(AccessorType& accessor,
+            const NearestNeighbors op)
+            : mOrigin(nullptr)
+            , mNeighbours(NodeMaskOp::ksize(op), nullptr)
+            , mAccessor(&accessor)
+            , mOnTile(true)
+            , mOffTile(false)
+            , mOp(op) {}
+
+        /// @brief Dilate a single leaf node by the current spatial scheme
+        ///        stored on the instance of this NodeMaskOp. Neighbour leaf
+        ///        nodes are also updated.
+        /// @details  Unlike erode, dilate is expected to be called in a
+        ///           single threaded context as it will update the node masks
+        ///           of neighbouring leaf nodes as well as the provided leaf.
+        /// @param  leaf  The leaf to dilate. The leaf's origin and value mask
+        ///               are used to calculate the result of the dilation.
+        inline void dilate(LeafType& leaf)
         {
-            if (init) {
-                init = false;
-                Coord orig = xyz.offsetBy(DX*LEAF_DIM, DY*LEAF_DIM, DZ*LEAF_DIM);
-                leaf = acc.probeLeaf(orig);
-                if ((leaf == nullptr) && !acc.isValueOn(orig)) leaf = acc.touchLeaf(orig);
-            }
-            static const int N = (LEAF_DIM - 1)*(DY + DX*LEAF_DIM);
-            if (leaf) leaf->getValueMask().template getWord<Word>(indx-N) |= mask;
+            // copy the mask
+            const MaskType cache = leaf.getValueMask();
```

Comment:
It confused me this was called `cache` not `mask`.

Reply:
renamed